Fix the bug when near filter is used

Amir-61 opened the pull request:
Fix the bug when `near` filter is used


* If near filter was used, users would get:

```
Error: Callback was already called
```

Connect to https://github.com/strongloop/loopback-connector-mongodb/issues/285

/to: @bajtos

Amir-61 commented:
@slnode test please

bajtos commented:
@Amir-61 please start by writing a test reproducing the issue.

TBH, the original code is a long spaghetti that I don't have capacity to grok right now. If all existing tests are passing and the new tests covering the bug is passing too, then I am ok to land this.

OTOH, if you can invest little bit more time, then it would be great to clean this code up first, before fixing the bug. For example, there are two places where `this.notifyObserversOf('access')` is called . Many of the `if/else` branches are too long and should be extracted into named functions too. And so on. The idea to is to clean up the code in such way that the commit fixing the bug will become a very small change.

Amir-61 commented:
@slnode test please

Amir-61 commented:
@bajtos 
I fixed more issues and added a test to verify the fixes; could you please review it?

>TBH, the original code is a long spaghetti

I agree, unfortunately that's the case for most of the code in DAO :(

>If all existing tests are passing and the new tests covering the bug is passing too, then I am ok to land this.

AFAICT all tests for juggler and connectors including `loopback-connector-mongodb` and others are passing now; also, I added test which was failing with the previous code base and passes with the changes here.

> there are two places where this.notifyObserversOf('access') is called 

That was the actual bug and I added test to verify the fix.

>Many of the if/else branches are too long and should be extracted into named functions too

I have named functions now `withNotifyGeo`, `queryGeo` and `allCb`, please let me know if you have more suggestions...

>The idea to is to clean up the code in such way that the commit fixing the bug will become a very small change.

The changes are not too much, but since I moved them around the GH changes are not easy to follow... sorry about that :(

Amir-61 commented on test/persistence-hooks.suite.js:
```diff
@@ -86,6 +86,18 @@ module.exports = function(dataSource, should, connectorCapabilities) {
+            { location: { near: '10,20', maxDistance: '10', unit: 'meters' }},
+          }, function(err, list) {
+          if (err) return done(err);
+          hookMonitor.names.should.eql(['access']);
```
Please also note I do not think we have ever had `loaded` hook if `near` filter was used. We may add it in another PR.

bajtos replied:
Is there any GH issue to keep track of this bug? Are you planning to fix this as part of  strongloop/loopback-connector-mongodb#285? Either way, please create a checklist item or a new GH issue so that we don't forget about this problem.

Amir-61 replied:
Create an issue and I assigned it to myself.

bajtos commented on test/persistence-hooks.suite.js:
```diff
@@ -86,6 +86,18 @@ module.exports = function(dataSource, should, connectorCapabilities) {
           });
       });
 
+      it('should trigger access hook once, if near filter is used', function(done) {
```
How about the following name?

```js
it('triggers correct hooks when near filter is used', ...)
```

This way the test can be easily extended later when we fix the implementation to trigger the "loaded" hook too.

Amir-61 commented:
@rptally @bachirelkhoury @jainisenough @geodeveloper @jwebcat @Zvika88
Could one of you please verify that strongloop/loopback-connector-mongodb#285 is fixed by this patch juggler@2.x.

Thanks a lot!

Amir-61 commented:
@bajtos Could you please also sign off https://github.com/strongloop/loopback-datasource-juggler/pull/1125

Thanks a lot 🙏

vidhema commented:
> @bajtos Could you please also sign off #1125
> 
> To-do:
> 
> * [x]  release `loopback-datasource-juggler@2.53.0`
> 
> Thanks a lot

not working at all

bajtos commented:
@vidhema please open a new GitHub issue and provide us with enough information to allow us to reproduce the issue locally, see https://loopback.io/doc/en/contrib/Reporting-issues.html#loopback-3x-bugs
